### ISSUE TITLE
Evaluate if real-time multi-way classification is the bottleneck for fast-inference?

### ISSUE THREAD

**Issue by carbonscott** [carbonscott/deepprojection#11]

**Comment by carbonscott** [comment]
Not at all, the bottleneck is data loading.  Please refer to 955b987f4ab739d8b9e636a46ff7e85cac7067c3, 35f68a3b43f801dd867fd7003ef5a8dc2f38bbd5 and aa66214aaf1e46a6b48fdf6ef8717d36d11fccfd.

**Comment by carbonscott** [comment]
I profiled it manually using breakpoint and nvidia-smi.  I found memory significantly increases due to the image size (86 × 86) that leads to 13 Million parameters in my model.  Another interesting finding is that the memory is tripled when each of anchor, positive and negative is passed into the embedding model.  However, I suspect the slow training is not due to memory, because I found that training is still slow even when input images are resized down to 46 × 46.  My current guess is the data loading (maybe I use psana layer in an inefficient way)... I think I will try to cache 2000 images in cpu memory altogether before data loader moves a batch of them into GPU... hopefully it will solve the slow-training problem.

Later...

Caching solves the slow training.
